internal/debug/debug.go

```diff
@@ -0,0 +1,206 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package debug
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// LogLevelDebug will enable debugging for all applciations and SDKs. This
+	// uses log levels of firecracker_debug, firecracker-go-sdk,
+	// firecracker_output, and firecracker_containerd_debug
+	LogLevelDebug = "debug"
+	// LogLevelFirecrackerDebug enables debug level logging on Firecracker
+	// MicroVMs
+	LogLevelFirecrackerDebug = "firecracker_debug"
+	// LogLevelFirecrackerError enables error level logging on Firecracker
+	// MicroVMs
+	LogLevelFirecrackerError = "firecracker_error"
+	// LogLevelFirecrackerInfo enables info level logging on Firecracker MicroVMs
+	LogLevelFirecrackerInfo = "firecracker_info"
+	// LogLevelFirecrackerWarning enables warning level logging on Firecracker
+	// MicroVMs
+	LogLevelFirecrackerWarning = "firecracker_warning"
+	// LogLevelFirecrackerOutput will redirect stdout and stderr of Firecracker
+	// to logging. By default, stdout and stderr is ignored
+	LogLevelFirecrackerOutput = "firecracker_output"
+	// LogLevelFirecrackerSDK enables logging on firecracker-go-sdk. This will
+	// show what handlers and APIs are being called to Firecracker.
+	LogLevelFirecrackerSDK = "firecracker-go-sdk"
```

Comment:
The SDK uses Logrus and have log levels like others. Why don't we support that?

Reply:
Will add. Didn't realize that it also had log levels along with a debug boolean

Reply:
Ah nvm, the log level is for firecracker and to enable logging on the sdk it is just a boolean.

Reply:
For example, startVMM() has Printf, Errorf and Debugf. Is there a way to log SDK's Errof on firecracker-containerd's logs?

https://github.com/firecracker-microvm/firecracker-go-sdk/blob/18c53a97c81711965327fe00349a7f99dcd6e5d5/machine.go#L482

Reply:
Yea, we unfortunately don't expose a way of configuring that to the user D:

Reply:
Hmm. It is complicated than we thought. The SDK has

1. Debug flag
1. LogLevel
1. The way to pass Logrus's logger

The first one is only used by [client_transport.go](https://github.com/firecracker-microvm/firecracker-go-sdk/blob/a3f9b2f83cf947e6b3fc8da19a59c78b8e0c945e/client_transports.go#L44). [NewClient is taking the flag](https://github.com/firecracker-microvm/firecracker-go-sdk/blob/a3f9b2f83cf947e6b3fc8da19a59c78b8e0c945e/firecracker.go#L65). For clients that use Machine abstraction. That is passed from [cfg.Debug](https://github.com/firecracker-microvm/firecracker-go-sdk/blob/18c53a97c81711965327fe00349a7f99dcd6e5d5/machine.go#L338).

The second one is [Firecracker's level](https://github.com/firecracker-microvm/firecracker/blob/b3869fa29b500c08a021d161bb1cc669ec6a4ebf/src/api_server/swagger/firecracker.yaml#L661).

The third one is [used by our runtime](https://github.com/firecracker-microvm/firecracker-containerd/blob/1e7b50ed4b41575f88afe7c3200a5fe698b773ab/runtime/service.go#L515).

Reply:
So setting firecracker-containerd_info affects the SDK since the runtime is passing the logger to SDK as is. I'm unsure what is the right way to fix that. Having a separate logger as we wrote in the doc?

Reply:
Ah, yea, completely forgot you could pass the logger in yourself. Makes this way more complicated than it needs to be ><. Hm, I will go ahead and add the `firecracker-go-sdk_XX` levels and figure out a good way to pass the logger. I think I'll just have the `Helper` construct one and a getter to retrieve it.

I will spend a little time thinking about it and see if I can come up with something better